regression: unable to view reported messages of deleted users

Dnouv opened the pull request:
<!-- This is a pull request template, you do not need to uncomment or remove the comments, they won't show up in the PR text. -->

<!-- Your Pull Request name should start with one of the following tags
  feat: Adding a new feature
  refactor: A code change that doesn't change behavior (it doesn't add anything and doesn't fix anything)
  fix: For bug fixes that affect the end-user
  chore: For small tasks
  docs: For documentation
  ci: For updating CI configuration
  test: For adding tests
  i18n: For updating any translations
  regression: Issues created/reported/fixed during the development phase. kind of problem that never existed in production and that we don't need to list in a changelog for the end user
-->

If a user deletes their own account or if the user account is deleted. Then, on the Moderation Console the Admin was not able to view the reported messages.
This PR introduces a fix for this issue.
<!-- Checklist!!! If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code. 
  - I have read the Contributing Guide - https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat doc
  - I have signed the CLA - https://cla-assistant.io/RocketChat/Rocket.Chat
  - Lint and unit tests pass locally with my changes
  - I have added tests that prove my fix is effective or that my feature works (if applicable)
  - I have added necessary documentation (if applicable)
  - Any dependent changes have been merged and published in downstream modules
-->

## Proposed changes (including videos or screenshots)
<!-- CHANGELOG -->
<!--
  Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
  If it fixes a bug or resolves a feature request, be sure to link to that issue below.
  This description will appear in the release notes if we accept the contribution.
-->

If the "Delete all messages" is selected under the `Message Erasure` (`Setting>Messages`) setting, then if a user account is deleted, all the reported messages will be hidden, and they no longer show up on the Moderation Console.

If any other option is selected in the setting `Message Erasure`, then the reported messages will now show up in the Moderation Console, but there will be an additional Callout component, and some moderation actions against the user will be disabled since they are not possible to take on a deleted user.
<img width="250" src="https://github.com/RocketChat/Rocket.Chat/assets/61188295/59b6800f-f54d-4ca4-8eef-1e1f4071add7" />

> Why not delete all the reports against a message when the message is deleted?

Eren got trolled by a random user, Eren got sad, and he reported those messages, expecting to get some justice, but soon the troller deleted all those messages (now the reports will also be dismissed/hidden), and here Eren is left waiting for Moderator to take some action on this troller 😟 .


## Why introduce a new method

Before this PR is merged, there could be some edge cases where the user is already deleted along with all messages, and this will cause an error while deleting their messages (as these messages no longer exist in the `messages` collection but do show up in the `moderation_reports` collection because, at the time of reporting a message, we keep a snapshot of the message) from the Moderation Console.
The error occurs because before the `deleteReportedMessage` used to call the `Messages` model method `cloneAndSaveAsHistoryById`, which as written [here](https://github.com/RocketChat/Rocket.Chat/blob/c10137e015b2cfa42aa7976d88e512e2ed0ed34e/apps/meteor/server/models/raw/Messages.ts#L1011-L1012) returns an error if a message is not found. 

To tackle this issue, my first approach was inside the `deleteReportedMessages` function. I called the 
```ts
const existingMids = Messages.find({ _id: { $in: ids } }, {projection: {_id: 1}}).toArray() // where ids is an array of message-ids that are to be deleted

await Promise.all(existingMids.map((mid) => Messages.cloneAndSaveAsHistoryById(mid, user)))
```

however, the above approach loads all the messages into the memory, blocking the Event Loop. And after getting feedback from discussions with some more experienced members, we found a better way of doing this by introducing a new method, which takes in the `message` as an argument.

Advantages of the new method.
- Reduced the DB call by 50%; before, it used to be 2 trips per message to delete, now it's only 1.
- No need to load all messages into memory since the Mongo cursor loads the documents in batches.

this returns the messages
<!-- END CHANGELOG -->

## Issue(s)
<!-- Link the issues being closed by or related to this PR. For example, you can use #594 if this PR closes issue number 594 -->

## Steps to test or reproduce
<!-- Mention how you would reproduce the bug if not mentioned on the issue page already. Also mention which screens are going to have the changes if applicable -->

1. Create a temporary user (temp).
2. Log in as the temp user, go into any channel and send a message, and report this message.
3. Go to “My Account“ and click “Delete My Account“
4. Head over to the Moderation Console from an Admin account and try clicking on the row of the “temp” user.
5. The Context bar will remain blank and an error shows up `invalid-user`

## Further comments
<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

[WM-120](https://rocketchat.atlassian.net/browse/WM-120)

[WM-120]: https://rocketchat.atlassian.net/browse/WM-120?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

debdutdeb commented on apps/meteor/app/api/server/v1/moderation.ts:
```diff
@@ -73,9 +73,12 @@ API.v1.addRoute(
-			const user = await Users.findOneById(userId, { projection: { _id: 1 } });
-			if (!user) {
-				return API.v1.failure('error-invalid-user');
+			const existingReport = await ModerationReports.findOne(
+				{ 'message.u._id': userId, '_hidden': { $ne: true } },
+				{ projection: { _id: 1 } },
+			);
```
Move this to the model.

Dnouv replied:
Thank you! I will create a new method to handle this.

debdutdeb commented on apps/meteor/app/api/server/v1/moderation.ts:
```diff
@@ -73,9 +73,12 @@ API.v1.addRoute(
+				{ projection: { _id: 1 } },
+			);
+			if (!existingReport) {
+				return API.v1.failure('no-report-found-for-this-user');
```
```suggestion
				return API.v1.failure('no-report-found');
```

debdutdeb commented on apps/meteor/app/api/server/v1/moderation.ts:
```diff
@@ -126,9 +129,12 @@ API.v1.addRoute(
-			const user = await Users.findOneById(userId, { projection: { _id: 1 } });
-			if (!user) {
-				return API.v1.failure('error-invalid-user');
+			const existingReport = await ModerationReports.findOne(
+				{ 'message.u._id': userId, '_hidden': { $ne: true } },
+				{ projection: { _id: 1 } },
+			);
+			if (!existingReport) {
+				return API.v1.failure('no-report-found-for-this-user');
```
same suggestions as previous

debdutdeb commented on apps/meteor/app/lib/server/functions/deleteUser.ts:
```diff
@@ -25,7 +26,9 @@ export async function deleteUser(userId: string, confirmRelinquish = false): Pro
 	});
 
-	if (!user) {
+	const actionUserId = Meteor.userId();
+
+	if (!user || !actionUserId) {
```
pass the userid as an argument, avoid using meteor.

Dnouv replied:
That would require modifying the existing `deleteUser` function, and so potentially I will have to modify the files which call it, would that be ok?

Dnouv replied:
Here we need the uid of the user who is calling this function. Because while hiding the reported messages, the moderator's id is stored along with the action.

If a user deletes their own account, we can use the passed `user._id`, but if someone else calls the `deleteUser`, the userId of the user whose account is going to be deleted is passed on, and we don't need that. 

I hope that makes things clear. Thank you!

debdutdeb commented on apps/meteor/server/lib/moderation/deleteReportedMessages.ts:
```diff
@@ -11,15 +11,22 @@ export async function deleteReportedMessages(messages: IMessage[], user: IUser):
-	for (const message of messages) {
+
+	const promises = messages.map(async (message) => {
+		const foundMessage = await Messages.findOneById(message._id, { projection: { _id: 1 } });
```
I don't get this change. why do you have to check for every id?

debdutdeb reviewed (changes requested):
for `moderation.user.deleteReportedMessages` I'm wondering what happens if a user is deleted. Isn't all of their messages get deleted with them?

Dnouv commented:
Thank you for the review, @debdutdeb ! 

> for moderation.user.deleteReportedMessages I'm wondering what happens if a user is deleted. Isn't all of their messages get deleted with them?

Yes, you are partially correct; under the settings (`Settings>Messages>Message Erasure Type`), the WSpace Admin can configure what behavior they want, we provide three options:
1. **Keep Messages and User Name**: The message and files history of the user will be deleted from Direct Messages and will be kept in other rooms. 
2. **Delete All Messages**: All messages and files from the user will be deleted from the database and it will not be possible to locate the user anymore. 
3. **Remove link between user and messages**: This option will assign all messages and files of the user to Rocket.Cat bot and Direct Messages are going to be deleted.

And in this PR, I have added a few lines that will hide reports if option 2 is selected for the workspace and user account deletion is initiated.

And to manage the edge cases where the user was already deleted along with all of their messages before this PR was merged, I have modified the existing logic in `deleteReportedMessages` function. What it does is calls the `Messages.findOneById` and if the message is not found we don't take any action on that.

Thank you!

debdutdeb commented:
> And in this PR, I have added a few lines that will hide reports if option 2 is selected for the workspace and user account deletion is initiated.

That's a good catch. But imo should be done for all types. How a message is deleted shouldn't matter - if it's deleted it's not viewabl thus reports should be dismissed. 

> And to manage the edge cases where the user was already deleted along with all of their messages before this PR was merged, I have modified the existing logic in deleteReportedMessages function. What it does is calls the Messages.findOneById and if the message is not found we don't take any action on that.

This part I dislike. It's a lot of roundtrips. What happens if an id is invalid in the following operations? Aren't they ignored (deleteMany, updateMany)?

Dnouv commented:
> This part I dislike. It's a lot of roundtrips. What happens if an id is invalid in the following operations? Aren't they ignored (deleteMany, updateMany)?

I don't know why, but this was the first time I saw an error thrown inside a Model 

https://github.com/Dnouv/Rocket.Chat/blob/1424fa8bd64e684f36ffe59c23af574b25ea0e82/apps/meteor/server/models/raw/Messages.ts#L1011

Dnouv commented:
> But imo should be done for all types. How a message is deleted shouldn't matter - if it's deleted it's not viewabl thus reports should be dismissed.

No, actually, sorry but I personally think we should not do this, I thought from a Moderator's POV:

Eren got trolled by a random user, Eren got sad, and he reported those messages, expecting to get some justice, but soon the troller deletes all those messages, and here Eren is left waiting for Moderator to take some action on this troller 😟 .

debdutdeb commented on apps/meteor/server/lib/moderation/deleteReportedMessages.ts:
```diff
@@ -20,9 +21,15 @@ export async function deleteReportedMessages(messages: IMessage[], user: IUser):
 	if (keepHistory) {
 		if (showDeletedStatus) {
-			await Promise.all(messageIds.map((id) => Messages.cloneAndSaveAsHistoryById(id, user as any)));
+			const existingMessages = await Messages.find({ _id: { $in: messageIds } }).toArray();
+
+			if (existingMessages.length === 0) {
+				return;
+			}
+			await Promise.all(existingMessages.map((msg) => Messages.cloneAndSaveAsHistoryById(msg._id, user as any)));
```
adding a new method won't change existing behaviour.

debdutdeb replied:
actually nvm no need for another method. my bad

KevLehman commented:
Isn't this like too big for a regression? 👀

Dnouv commented:
Is it? There is a fix in this PR related to i18n (the WM-143), which was classified as a regression.
